### PR TITLE
Allowing CylindricalMesh and SphericalMesh to be fully made via constructor+ type hints

### PR DESCRIPTION
**Pull request by shimwell**
# Description

I've been crawling through the code looking for classes that are not fully setable from their args and I found CylindricalMesh and SphericalMesh were just missing the origin arg.

So I've added that argument to their __ init __

While browsing the code I couldn't resist adding a few type hints

Fixes # (issue)

# Checklist

- [x] I have performed a self-review of my own code
<s> - [ ] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) on any C++ source files (if applicable) </s>
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

File: openmc/mesh.py [hunk 1]
```diff
@@ -1204,22 +1207,24 @@ class CylindricalMesh(StructuredMesh):
...
         An iterable of mesh indices for each mesh element, e.g. [(1, 1, 1),
         (2, 1, 1), ...]
 
     """
 
-    def __init__(self, mesh_id: int = None, name: str = ''):
+    def __init__(
+        self,
+        mesh_id: int = None,
+        name: str = '',
```
Comment (eepeterson) [inline openmc/mesh.py:1219]: need to add `r_grid`, `phi_grid`, and `z_grid` as parameters here as well.

File: openmc/mesh.py [hunk 2]
```diff
@@ -1505,22 +1513,24 @@ class SphericalMesh(StructuredMesh):
...
         (2, 1, 1), ...]
 
     """
 
-    def __init__(self, mesh_id=None, name=''):
+    def __init__(
+        self,
+        mesh_id: typing.Optional[int] = None,
+        name: str = '',
+        origin: typing.Tuple[float] = (0, 0, 0)
```
Comment (eepeterson) [inline openmc/mesh.py:1526]: same comment as above: need to be able to set the grids in the constructor.

File: openmc/mesh.py [hunk 3]
```diff
@@ -1514,13 +1542,21 @@ class SphericalMesh(StructuredMesh):
 
     """
 
-    def __init__(self, mesh_id=None, name=''):
+    def __init__(
+        self,
+        mesh_id: typing.Optional[int] = None,
+        name: str = '',
+        origin: typing.Tuple[float] = (0, 0, 0),
```
Comment (shimwell) [inline openmc/mesh.py:1549]:
```suggestion
        origin: typing.Tuple[float] = (0., 0., 0.),
```

File: openmc/mesh.py [hunk 4]
```diff
@@ -1204,22 +1215,27 @@ class CylindricalMesh(StructuredMesh):
...
         (2, 1, 1), ...]
 
     """
 
-    def __init__(self, mesh_id: int = None, name: str = ''):
+    def __init__(
+        self,
+        mesh_id: int = None,
+        name: str = '',
+        origin: typing.Tuple[float] = (0., 0., 0.),
```
Comment (shimwell) [inline openmc/mesh.py:1228]: Just to mention I've added a ```Tuple``` type hint to origin as that agrees with the default value used. However the doc string asks for an array and I notice we change it to an array in the setter. Not sure what is preferable here but I guess these are the sorts of things we uncover while adding type hints

File: openmc/mesh.py [hunk 5]
```diff
@@ -1514,13 +1542,21 @@ class SphericalMesh(StructuredMesh):
 
     """
 
-    def __init__(self, mesh_id=None, name=''):
+    def __init__(
+        self,
+        mesh_id: typing.Optional[int] = None,
+        name: str = '',
+        origin: typing.Tuple[float] = (0., 0., 0.),
```
Comment (shimwell) [inline openmc/mesh.py:1549]: Same for the type hinting on this origin as well, perhaps we should have the default values as an np.array([0, 0, 0]) and the type hint as an np.array

File: openmc/mesh.py (hunk 4 above)
Comment (eepeterson) [reply openmc/mesh.py:1228]: I think it’s preferable to have the type hint be an iterable of float or real if that’s possible and for the internal representations and what you get back from getters to be numpy arrays

File: openmc/stats/univariate.py [hunk 6]
```diff
@@ -1,19 +1,20 @@
...
-import math
 from numbers import Real
 from warnings import warn
-import lxml.etree as ET
 
+import lxml.etree as ET
 import numpy as np
 
 import openmc.checkvalue as cv
+
```
Comment (paulromano) [inline openmc/stats/univariate.py:14]: These imports should still be together
```suggestion
```

File: tests/unit_tests/test_mesh.py [hunk 7]
```diff
@@ -1,6 +1,10 @@
-import openmc
-import pytest
+from math import pi
+
 import numpy as np
+import pytest
+
+import openmc
```
Comment (paulromano) [inline tests/unit_tests/test_mesh.py:6]: From the perspective of tests, `openmc` is a third-party package, not the local package
```suggestion
import pytest
import openmc
```

**Review by paulromano — CHANGES_REQUESTED** [review]
Thanks for this PR @shimwell!

File: openmc/stats/multivariate.py [hunk 8]
```diff
@@ -209,7 +211,7 @@ class Monodirectional(UnitSphere):
 
     """
 
-    def __init__(self, reference_uvw=[1., 0., 0.]):
+    def __init__(self, reference_uvw: typing.Sequence[float]=[1., 0., 0.]):
```
[the remark this answers is not in the document]
Comment (shimwell) [reply openmc/stats/multivariate.py:214]: I remember we had a conflict for Union and typing.Union so I shall avoid that one but all the others I shall get rid of the ```typing.``` part

File: tests/unit_tests/test_mesh.py (hunk 7 above)
Comment (shimwell) [reply tests/unit_tests/test_mesh.py:6]: Thanks Paul, I trusted [isort](https://pycqa.github.io/isort/) a bit much here

File: openmc/mesh.py [hunk 9]
```diff
@@ -1204,22 +1215,27 @@ class CylindricalMesh(StructuredMesh):
...
 
-    def __init__(self, mesh_id: int = None, name: str = ''):
+    def __init__(
+        self,
+        mesh_id: int = None,
+        name: str = '',
+        origin: typing.Tuple[float] = (0., 0., 0.),
+        r_grid: typing.Optional[np.ndarray] = None,
+        phi_grid: np.ndarray = np.array([0, 2*pi]),
+        z_grid: typing.Optional[np.ndarray] = None,
```
[the remark this answers is not in the document]
Comment (eepeterson) [reply openmc/mesh.py:1231]: > As long as we're mucking around with the arguments, I personally think we should do things right and move things that are effectively "required" to the front of the argument list and give them no defaults. In this case, there's no reasonable default for `r_grid` so the user should always need to specify that. So, I think a reasonable order of arguments is `r_grid`, `phi_grid`, `z_grid`, and then everything else.

Are you also suggesting moving the `mesh_id` and `name` kwargs to the back of the list? That would be great if we could do that and start propagating things throughout the API although it is largely breaking if people are using those args as positional arguments... (which I think we should discourage anyway).

[the remark this answers is not in the document]
Comment (paulromano) [reply openmc/mesh.py:1231]: Correct, move 'em to the back. I fully realize changing argument order will break things for some but think it's totally worth it, especially for these classes which likely haven't seen a _ton_ of use yet. And yes, I discourage folks from manually specifying IDs at every chance I get!

File: openmc/mesh.py [hunk 10]
```diff
@@ -1,19 +1,21 @@
...
 from abc import ABC, abstractmethod
 from collections.abc import Iterable
 from math import pi
-from numbers import Real, Integral
+from numbers import Integral, Real
 from pathlib import Path
-import typing
-import warnings
-import lxml.etree as ET
+from typing import Iterable, Optional, Sequence, Tuple
```
Comment (paulromano) [inline openmc/mesh.py:8]: We also import `Iterable` from `collections.abc`; one of them needs to change (maybe better to explicitly type `collections.abc.Iterable` and use `Iterable` to mean `typing.Iterable`)

File: openmc/mesh.py [hunk 11]
```diff
@@ -1,19 +1,21 @@
...
 import h5py
+import lxml.etree as ET
 import numpy as np
 
-import openmc.checkvalue as cv
 import openmc
-from ._xml import get_text
+import openmc.checkvalue as cv
 from openmc.checkvalue import PathLike
+
```
Comment (paulromano) [inline openmc/mesh.py:17]:
```suggestion
```

File: openmc/mesh.py [hunk 12]
```diff
@@ -1223,12 +1225,12 @@ class CylindricalMesh(StructuredMesh):
 
     def __init__(
         self,
+        origin: Tuple[float] = (0., 0., 0.),
+        r_grid: Sequence[float] = (0, 10),
```
Comment (paulromano) [inline openmc/mesh.py:1229]: Again, I don't think it makes sense to assign a default `r_grid` and would prefer that this be a required argument with no default. In this case, it seems cleaner to have the three grids first, then origin, then everything else.

File: openmc/mesh.py [hunk 13]
```diff
@@ -1544,12 +1546,12 @@ class SphericalMesh(StructuredMesh):
 
     def __init__(
         self,
-        mesh_id: typing.Optional[int] = None,
+        origin: Tuple[float] = (0., 0., 0.),
+        r_grid: Sequence[float] = (0, 10),
```
Comment (paulromano) [inline openmc/mesh.py:1550]: Same thing re: grids and origin

**Review by paulromano — CHANGES_REQUESTED** [review]

File: openmc/mesh.py (hunk 13 above)
Comment (shimwell) [reply openmc/mesh.py:1550]: Thanks Paul, how about z_grid?

Comment (paulromano) [reply openmc/mesh.py]: Right, z_grid also doesn't have a sensible default in my opinion, so we should make that required as well

File: openmc/mesh.py [hunk 14]
```diff
@@ -1830,7 +1872,7 @@ def volumes(self):
         return self._volumes
 
     @volumes.setter
-    def volumes(self, volumes: typing.Iterable[Real]):
+    def volumes(self, volumes: Iterable[Real]):
```
Comment (shimwell) [inline openmc/mesh.py:1875]:
```suggestion
    def volumes(self, volumes: typing.Iterable[Real]):
```

File: openmc/mesh.py [hunk 15]
```diff
@@ -1213,13 +1225,21 @@ class CylindricalMesh(StructuredMesh):
 
     """
 
-    def __init__(self, mesh_id: int = None, name: str = ''):
+    def __init__(
+        self,
+        r_grid: Sequence[float],
+        z_grid: Sequence[float],
+        phi_grid: Sequence[float] = (0, 2*pi),
+        origin: Tuple[float] = (0., 0., 0.),
```
Comment (paulromano) [inline openmc/mesh.py:1233]:
```suggestion
        origin: Sequence[float] = (0., 0., 0.),
```

File: openmc/mesh.py [hunk 16]
```diff
@@ -1309,7 +1329,7 @@ def from_hdf5(cls, group: h5py.Group):
         mesh_id = int(group.name.split('/')[-1].lstrip('mesh '))
 
         # Read and assign mesh properties
-        mesh = cls(mesh_id)
+        mesh = cls(mesh_id=mesh_id)
```
Comment (paulromano) [inline openmc/mesh.py:1332]: This is not going to work since `r_grid` and `z_grid` are now required arguments

File: openmc/mesh.py [hunk 17]
```diff
@@ -1213,13 +1225,21 @@ class CylindricalMesh(StructuredMesh):
...
     """
 
-    def __init__(self, mesh_id: int = None, name: str = ''):
+    def __init__(
+        self,
+        r_grid: Sequence[float],
+        z_grid: Sequence[float],
+        phi_grid: Sequence[float] = (0, 2*pi),
+        origin: Tuple[float] = (0., 0., 0.),
+        mesh_id: int = None,
```
Comment (paulromano) [inline openmc/mesh.py:1234]:
```suggestion
        mesh_id: Optional[int] = None,
```

File: openmc/mesh.py [hunk 18]
```diff
@@ -1514,13 +1548,21 @@ class SphericalMesh(StructuredMesh):
 
     """
 
-    def __init__(self, mesh_id=None, name=''):
+    def __init__(
+        self,
+        r_grid: Sequence[float],
+        phi_grid: Sequence[float] = (0, 2*pi),
+        theta_grid: Sequence[float] = (0, pi),
+        origin: Tuple[float] = (0., 0., 0.),
```
Comment (paulromano) [inline openmc/mesh.py:1556]:
```suggestion
        origin: Sequence[float] = (0., 0., 0.),
```

File: openmc/mesh.py [hunk 19]
```diff
@@ -1610,7 +1652,7 @@ def from_hdf5(cls, group: h5py.Group):
         mesh_id = int(group.name.split('/')[-1].lstrip('mesh '))
 
         # Read and assign mesh properties
-        mesh = cls(mesh_id)
+        mesh = cls(mesh_id=mesh_id)
```
Comment (paulromano) [inline openmc/mesh.py:1655]: Won't work since there are now required arguments

File: openmc/mesh.py [hunk 20]
```diff
@@ -1663,7 +1705,7 @@ def from_xml_element(cls, elem: ET.Element):
 
         """
         mesh_id = int(get_text(elem, 'id'))
-        mesh = cls(mesh_id)
+        mesh = cls(mesh_id=mesh_id)
```
Comment (paulromano) [inline openmc/mesh.py:1708]: Won't work since there are now required arguments

**Review by paulromano — CHANGES_REQUESTED** [review]
In addition to line comments below, it looks like more changes were now made in other files (not relevant to original purpose of PR) -- I'll review those but I'd appreciate if we can limit the scope of the PR in the future

File: openmc/stats/multivariate.py [hunk 21]
```diff
@@ -1,15 +1,19 @@
+from __future__ import annotations
+
```
Comment (paulromano) [inline openmc/stats/multivariate.py:2]:
```suggestion
```

File: openmc/stats/multivariate.py [hunk 22]
```diff
@@ -1,15 +1,19 @@
...
-from math import pi, cos
+from math import cos, pi
 from numbers import Real
-import lxml.etree as ET
 
+import lxml.etree as ET
 import numpy as np
 
 import openmc.checkvalue as cv
+
```
Comment (paulromano) [inline openmc/stats/multivariate.py:13]:
```suggestion
```

File: openmc/stats/univariate.py [hunk 23]
```diff
@@ -793,8 +797,13 @@ class Tabular(Univariate):
 
     """
 
-    def __init__(self, x, p, interpolation='linear-linear',
-                 ignore_negative=False):
+    def __init__(
+            self,
+            x: np.ndarray,
+            p: np.ndarray,
```
Comment (paulromano) [inline openmc/stats/univariate.py:803]:
```suggestion
            x: typing.Sequence[float],
            p: typing.Sequence[float],
```

File: tests/unit_tests/test_mesh.py [hunk 24]
```diff
@@ -37,10 +40,63 @@ def test_raises_error_when_flat(val_left, val_right):
...
     bb = mesh.bounding_box
     assert isinstance(bb, openmc.BoundingBox)
     np.testing.assert_array_equal(bb.lower_left, np.array([-2, -3 ,-5]))
     np.testing.assert_array_equal(bb.upper_right, np.array([2, 3, 5]))
 
+
+def test_SphericalMesh_initiation():
+
+    # test defaults
+    mesh = openmc.SphericalMesh(r_grid = (0, 10))
```
Comment (paulromano) [inline tests/unit_tests/test_mesh.py:54]:
```suggestion
    mesh = openmc.SphericalMesh(r_grid=(0, 10))
```

File: tests/unit_tests/test_mesh.py [hunk 25]
```diff
@@ -37,10 +40,63 @@ def test_raises_error_when_flat(val_left, val_right):
...
+
+    # test attribute changing
+    mesh.r_grid = (0, 11)
+    assert (mesh.r_grid == np.array([0., 11.])).all()
+
+
+def test_CylindricalMesh_initiation():
+
+    # test defaults
+    mesh = openmc.CylindricalMesh(r_grid = (0, 10), z_grid = (0, 10))
```
Comment (paulromano) [inline tests/unit_tests/test_mesh.py:80]:
```suggestion
    mesh = openmc.CylindricalMesh(r_grid=(0, 10), z_grid=(0, 10))
```

**Review by paulromano — CHANGES_REQUESTED** [review]

File: tests/unit_tests/test_filter_mesh.py [hunk 26]
```diff
@@ -75,17 +77,19 @@ def test_cylindrical_mesh_estimators(run_in_tmpdir):
     model.settings.inactive = 10
     model.settings.batches = 20
 
-    cyl_mesh = openmc.CylindricalMesh()
-    cyl_mesh.r_grid = np.linspace(0.0, 5.0**3, 20)**(1/3)
-    cyl_mesh.z_grid = [-5., 5.]
+    cyl_mesh = openmc.CylindricalMesh(
+        r_grid = np.linspace(0.0, 5.0**3, 20)**(1/3),
+        z_grid = [-5., 5.]
```
Comment (shimwell) [inline tests/unit_tests/test_filter_mesh.py:82]:
```suggestion
        r_grid=np.linspace(0.0, 5.0**3, 20)**(1/3),
        z_grid=[-5., 5.]
```

File: tests/unit_tests/test_filter_mesh.py [hunk 27]
```diff
@@ -75,17 +77,19 @@ def test_cylindrical_mesh_estimators(run_in_tmpdir):
...
     tally1.filters = [openmc.MeshFilter(cyl_mesh)]
     tally1.scores = ['flux']
     tally1.estimator = 'collision'
 
-    cyl_mesh = openmc.CylindricalMesh()
-    cyl_mesh.r_grid = np.linspace(0.0, 5.0**3, 20)**(1/3)
-    cyl_mesh.z_grid = [-5., 5.]
+    cyl_mesh = openmc.CylindricalMesh(
+        r_grid = np.linspace(0.0, 5.0**3, 20)**(1/3),
+        z_grid = [-5., 5.]
```
Comment (shimwell) [inline tests/unit_tests/test_filter_mesh.py:91]:
```suggestion
        r_grid=np.linspace(0.0, 5.0**3, 20)**(1/3),
        z_grid=[-5., 5.]
```

File: tests/unit_tests/test_filter_mesh.py [hunk 28]
```diff
@@ -133,10 +137,11 @@ def test_cylindrical_mesh_coincident(scale, run_in_tmpdir):
...
     model.settings.inactive = 0
 
-    cyl_mesh = openmc.CylindricalMesh()
-    cyl_mesh.r_grid = [0., 1.25*scale]
-    cyl_mesh.phi_grid = [0., 2*math.pi]
-    cyl_mesh.z_grid = [-1e10, 1e10]
+    cyl_mesh = openmc.CylindricalMesh(
+        r_grid = [0., 1.25*scale],
+        phi_grid = [0., 2*math.pi],
+        z_grid = [-1e10, 1e10]
```
Comment (shimwell) [inline tests/unit_tests/test_filter_mesh.py:143]:
```suggestion
        r_grid=[0., 1.25*scale],
        phi_grid=[0., 2*math.pi],
        z_grid=[-1e10, 1e10]
```

**Comment by shimwell** [comment]
I did have to tinker with a few extra test files that were calling openmc.CylindricalMesh() and openmc.SphericalMesh() without arguments (which is no longer not possible as we now have mandatory args) but looks like the CI is now passing :tada:

**Review by paulromano — APPROVED** [review]
Sorry for the delay on this -- thanks for the updates @shimwell!

**Comment by shimwell** [comment]
Many thanks Paul, sorry it was not the most tidy PR.

**Pull request merged**